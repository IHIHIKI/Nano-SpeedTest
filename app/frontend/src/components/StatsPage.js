import React, {Component} from 'react';
import Header from './Header';
import TableRow from './TableRow';
import '../styles/StatsPage.css';
import Map from './Map';

class StatsPage extends Component {
    state = {
        loading: true,
        fakeTableData: [
            {
                hash: 'abc-123',
                origin: 'SF',
                destination: 'LA',
                time: 9.3,
                amount: 12345,
                completed: 'Jan 1, 2018'
            }, {
                hash: 'zyx-321',
                origin: 'NY',
                destination: 'WI',
                time: 9.7,
                amount: 55555,
                completed: 'March 15, 2015'
            }
        ],
        cities: [{
            name: 'japan',
            coords: {
                lat: 36.2048,
                lng: 138.2529
            }
        },
        {
            name: 'sf',
            coords: {
                lat: 37.768249,
                lng: -122.445145
            }
        }
    ]
    };

    getData = () => {
        setTimeout(() => {
            this.setState({
                fakeTableData: [...this.state.fakeTableData,
                    {
                        hash: 'ahd-949',
                        origin: 'Madison',
                        destination: 'Austin',
                        time: 4.3,
                        amount: 54321,
                        completed: 'Sept 19, 2018'
                    }],
                loading: false
            });
        }, 4300)
    };

    handleClick = () => {
        this.setState({
            fakeTableData: [...this.state.fakeTableData.slice(0, this.state.fakeTableData.length - 1)],
            loading: true
        });
    };

    render() {
        // call function to "get data" once (set timer and then add value to array) and to simulate a single
        // async request
        if (this.state.loading) {
            this.getData();
        }

        // render the jsx
        return (
            <div className='StatsPage'>
                <Header/>
                <h1 className="page-header text-center">Stats Page</h1>
                <Map cities={this.state.cities}/>
                <table className="table">
                    <thead>
                    <tr>
                        <th scope='col'>#</th>
                        <th scope='col'>Hash</th>
                        <th scope='col'>Origin</th>
                        <th scope='col'>Destination</th>
                        <th scope='col'>Elapsed Time</th>
                        <th scope='col'>Amount</th>
                        <th scope='col'>Completed</th>
                    </tr>
                    </thead>
                    <tbody>
                    {
                        this.state.fakeTableData.map((props, index) => {
                            props.index = index + 1;
                            return <TableRow key={props.hash} {...props}/>;
                        })
                    }
                    </tbody>
                </table>

                {/*only show loading if we are waiting for the data.  Else show button to rerun*/}
                {this.state.loading ? (
                    <div className='loading-container'>
                        <div className='loader-container d-flex justify-content-center'>
                            <div className='loader'></div>
                        </div>
                        <div>
                            <p>Your transaction is processing. Please wait.</p>
                        </div>
                    </div> ) : (
                    <div className='loader-container d-flex justify-content-center'>
                        <button className='btn btn-primary' onClick={this.handleClick}>Run Test Again</button>
                    </div>
                )}
            </div>
        );
    }
}

export default StatsPage;