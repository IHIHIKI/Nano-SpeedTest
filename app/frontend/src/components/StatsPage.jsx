import React, {Component} from 'react';
import Header from './Header';
import PropTypes from 'prop-types';
import '../styles/StatsPage.css';
import Map from './Map';
import Table from './Table';
import NoTableEntries from './NoTableEntries';
import {connect} from 'react-redux';
import Ad from './Ad';

const loader = (
    <div className='loading-container'>
    <div className='loader-container d-flex justify-content-center'>
        <div className='loader'/>
    </div>
    <div>
        <p>Your transaction is processing. Please wait.</p>
    </div>
</div>
);


class StatsPage extends Component {
    state = {

    };

    render() {
        const {table, isFetchingTiming, isFetchingTransaction} = this.props;
        const mostRecent = table.length && table[table.length - 1];
        const sendMessage = mostRecent.completed ? 'Sent' : 'Sending';
        // render the jsx
        return (
            <div className='StatsPage'>
                <Header/>
                <Ad/>
                {(mostRecent && mostRecent.error && !isFetchingTransaction) ?
                    <div className='alert alert-danger' role='alert'>
                        Something went wrong while trying to get the transaction. Please try again
                    </div>
                    : null
                }
                { (isFetchingTransaction || mostRecent || isFetchingTiming) ? ( // this is pretty ugly and should be refactored in V2
                    <>{
                        (isFetchingTransaction || !mostRecent) ? loader // show loader if no data or is getting a transaction
                        : (<>
                            <h2 className='map-header page-header text-left'>
                            {sendMessage} from {mostRecent.origin.nodeLocation} to {mostRecent.destination.nodeLocation}
                            { isFetchingTiming && <>
                                <p/>
                                This can take upwards of 30 seconds so please be patient!
                            </>}
                            </h2>
                            <div className='nano-container'>
                                <Map {...mostRecent}/>
                            </div>
                            <Table tableData={table}/>
                        </>)
                    }</>
                    ) : <NoTableEntries /> // show message to run a test first
                }
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        table: state.table,
        isFetchingTransaction: state.transactions.isFetchingTransaction,
        isFetchingTiming: state.transactions.isFetchingTiming
    };
};


StatsPage.propTypes = {
    table: PropTypes.array.isRequired,
    isFetchingTransaction: PropTypes.bool,
    isFetchingTiming: PropTypes.bool
};

export default connect(mapStateToProps)(StatsPage);