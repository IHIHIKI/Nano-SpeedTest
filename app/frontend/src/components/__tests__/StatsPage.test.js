import React from 'react';
import renderer from 'react-test-renderer';
import StatsPage from '../StatsPage';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { BrowserRouter} from 'react-router-dom';
import { shallow } from 'enzyme';

const mockStore = configureStore();
let store = mockStore({
    navigation: {
        activeTab: ''
    },
    table: [
        {
            id: 148,
            origin: {
              id: 2,
              nodeLocation: 'Mumbai',
              latitude: '19.014400',
              longitude: '72.847900',
              coords: {
                lat: 19.0144,
                lng: 72.8479
              }
            },
            destination: {
              id: 1,
              nodeLocation: 'N. Virginia',
              latitude: '39.043700',
              longitude: '-77.487500',
              coords: {
                lat: 39.0437,
                lng: -77.4875
              }
            },
            amount: 0.0009,
            ip: '127.0.0.1',
            completed: false,
            index: 1
          }
    ],
    transactions: {

    }
});

const props = {
	history: {}
};

it('shallow renders without crashing', () => {
    shallow(<StatsPage store={store} {...props}/>);
});

it(`renders snapshot correctly`, () => {
    const tree = renderer.create(
        <Provider store={store}>
            <BrowserRouter >
                <StatsPage {...props}/>
            </BrowserRouter>
        </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});
