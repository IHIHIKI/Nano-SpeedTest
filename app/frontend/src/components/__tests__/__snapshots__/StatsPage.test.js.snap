// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot correctly 1`] = `
<div
  className="StatsPage"
>
  <div
    className="nav nav-tabs"
  >
    <li
      className="nav-item"
    >
      <a
        className="nav-link active"
        href="/"
        onClick={[Function]}
      >
         
        Speed Test
         
      </a>
    </li>
    <li
      className="nav-item"
    >
      <a
        className="nav-link "
        href="/Stats"
        onClick={[Function]}
      >
         
        Statistics
         
      </a>
    </li>
    <li
      className="nav-item"
    >
      <a
        className="nav-link "
        href="/Info"
        onClick={[Function]}
      >
         
        More Info
         
      </a>
    </li>
  </div>
  <h2
    className="map-header page-header text-left"
  >
    Sending
     from 
    Mumbai
     to 
    N. Virginia
  </h2>
  <div
    className="nano-container"
  >
    <div
      className="GoogleMapsContainer"
    >
      <div
        className="GoogleMapsElement"
      />
      <div />
    </div>
  </div>
  <div
    className="nano-container"
  >
    <table
      className="table"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            #
          </th>
          <th
            scope="col"
          >
            Id
          </th>
          <th
            scope="col"
          >
            Origin
          </th>
          <th
            scope="col"
          >
            Destination
          </th>
          <th
            scope="col"
          >
            Amount
          </th>
          <th
            scope="col"
          >
            Elapsed Time
          </th>
          <th
            scope="col"
          >
            Status
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className=""
        >
          <th
            scope="row"
          >
            1
          </th>
          <td>
            148
          </td>
          <td>
            Mumbai
          </td>
          <td>
            N. Virginia
          </td>
          <td>
            0.0009
             nano
          </td>
          <td>
            Pending...
          </td>
          <td>
            Pending...
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
